lib/set-visibility-hidden/SetVisibilityHidden.cpp

```diff
@@ -0,0 +1,62 @@
+#include <kllvm/codegen/SetVisibilityHidden.h>
+
+using namespace llvm;
+
+namespace kllvm {
+__attribute__((visibility("default"))) bool
+run_set_visibility_hidden(Module &m) {
+  bool dirty = false;
+  for (auto &global : m.globals()) {
+    if (!global.isDeclaration()) {
+      global.setVisibility(GlobalValue::HiddenVisibility);
+      dirty = true;
+    }
+  }
+  for (auto &func : m.functions()) {
+    if (!func.isDeclaration()) {
+      func.setVisibility(GlobalValue::HiddenVisibility);
+      dirty = true;
+    }
+  }
+  return dirty;
+}
+
+} // namespace kllvm
+
+using namespace kllvm;
+
+__attribute__((visibility("default"))) char legacy_set_visibility_hidden::ID
+    = 0;
+
+static RegisterPass<legacy_set_visibility_hidden>
+    x("set-visibility-hidden", "Set visibility of all global values to hidden",
+      false /* Only looks at CFG */, false /* Analysis Pass */);
+
+/* New PM Registration */
+llvm::PassPluginLibraryInfo get_set_visibility_hidden_plugin_info() {
+  return {
+      LLVM_PLUGIN_API_VERSION, "SetVisibilityHidden", LLVM_VERSION_STRING,
+      [](PassBuilder &pb) {
+        pb.registerPipelineStartEPCallback(
+            [](llvm::ModulePassManager &pm, OptimizationLevel level) {
+              pm.addPass(set_visibility_hidden());
+            });
+        pb.registerPipelineParsingCallback(
+            [](StringRef name, llvm::ModulePassManager &pm,
+               ArrayRef<llvm::PassBuilder::PipelineElement>) {
+              if (name == "set-visibility-hidden") {
+                pm.addPass(set_visibility_hidden());
+                return true;
+              }
+              return false;
+            });
+      }};
+}
+
+#ifndef LLVM_BYE_LINK_INTO_TOOLS
+extern "C" __attribute__((visibility("default")))
+LLVM_ATTRIBUTE_WEAK ::llvm::PassPluginLibraryInfo
+llvmGetPassPluginInfo() {
+  return get_set_visibility_hidden_plugin_info();
+}
+#endif
```

Comment:
It has been a while since the last time I wrote an LLVM Pass, so I would like to ask if it's really necessary to register both in the Legacy and in the New PassManager? Or is this a restriction from our source code?

Reply:
For the moment we need to use both because of the range of versions we support; I _think_ once we are able to fully upgrade to Ubuntu Noble then we'll be able to drop the requirement.

Reply:
That's good to know! Thanks!